File: backend/ts_template/devdonalds.ts
import express, { Request, Response } from "express";

// ==== Type Definitions, feel free to add or modify ==========================
interface cookbookEntry {
  name: string;
  type: string;
}

interface requiredItem {
  name: string;
  quantity: number;
}

interface recipe extends cookbookEntry {
  requiredItems: requiredItem[];
}

interface ingredient extends cookbookEntry {
  cookTime: number;
}

// =============================================================================
// ==== HTTP Endpoint Stubs ====================================================
// =============================================================================
const app = express();
app.use(express.json());

// Store your recipes here!
const cookbook: any = {};

// Task 1 helper (don't touch)
app.post("/parse", (req:Request, res:Response) => {
  const { input } = req.body;

  const parsed_string = parse_handwriting(input)
  if (parsed_string == null) {
    res.status(400).send("this string is cooked");
    return;
  } 
  res.json({ msg: parsed_string });
  return;
  
});

// [TASK 1] ====================================================================
// Takes in a recipeName and returns it in a form that 
const parse_handwriting = (recipeName: string): string | null => {
  recipeName = recipeName.trim()
                         .replace(/-/g, ' ')
                         .replace(/_/g, ' ')
                         .replace(/\s+/g, ' ')
                         .replace(/[^a-zA-Z\s]/g, '')
                         .toLowerCase()
                         .split(' ')
                         .map(word => word.charAt(0).toUpperCase() + word.slice(1))
                         .join(' ');
  return recipeName.length > 0 ? recipeName : null;
}

// [TASK 2] ====================================================================
// Endpoint that adds a CookbookEntry to your magical cookbook
app.post("/entry", (req:Request, res:Response) => {
  const data = req.body;
  if (data.type !== "recipe" && data.type !== "ingredient") {
    res.status(400).send("Wrong type");
    return;
  }

  if (data.type === "ingredient" && data.cookTime < 0) {
    res.status(400).send("Negative cook time");
    return;
  }

  if (data.name in cookbook) {
    res.status(400).send("Name not unique");
    return;
  }

  if (data.type === "recipe") {
    const seen = new Set<string>();
    const requiredItems: requiredItem[] = [];
    for (const item of data.requiredItems) {
      if (seen.has(item.name)) {
        res.status(400).send("Required item not unique");
        return;
      }
      seen.add(item.name);
      requiredItems.push({ name: item.name, quantity: item.quantity });
    }
    cookbook[data.name] = { ...data, requiredItems };
  } else {
    cookbook[data.name] = { ...data, cookTime: data.cookTime };
  }

  res.status(200).json({});
});

// [TASK 3] ====================================================================
// Endpoint that returns a summary of a recipe that corresponds to a query name
app.get("/summary", (req:Request, res:Request) => {
  const data = req.query.name;

  if (!data || !(data in cookbook)) {
    res.status(400).send("Recipe not found");
    return;
  }

  const root = cookbook[data];
  if (!("requiredItems" in root)) {
    res.status(400).send("Not a recipe");
    return;
  }

  const final: any = {
    name: root.name,
    cookTime: 0,
    ingredients: [],
    seen: new Set(),
  };

  if (!helper(root.requiredItems, 1, final)) {
    res.status(400).send("Ingredient not found");
    return;
  }

  res.status(200).json({
    name: final.name,
    cookTime: final.cookTime,
    ingredients: final.ingredients,
  });
});

const helper = (requiredItems: requiredItem[], multiplier: number, final: any): boolean => {
  for (const child of requiredItems) {
    if (!(child.name in cookbook)) {
      return false;
    }
    if ("cookTime" in cookbook[child.name]) {
      if (final.seen.has(child.name)) {
        const item = final.ingredients.find((i) => i.name === child.name);
        if (item) item.quantity += child.quantity * multiplier;
      } else {
        final.ingredients.push({ name: child.name, quantity: child.quantity * multiplier });
        final.seen.add(child.name);
      }
      final.cookTime += (cookbook[child.name] as ingredient).cookTime * child.quantity * multiplier;
    } else {
      if (!helper((cookbook[child.name] as recipe).requiredItems, child.quantity * multiplier, final)) {
        return false;
      }
    }
  }
  return true;
};

// =============================================================================
// ==== DO NOT TOUCH ===========================================================
// =============================================================================
const port = 8080;
app.listen(port, () => {
  console.log(`Running on: http://127.0.0.1:8080`);
});

File: backend/py_template/devdonalds.py
from dataclasses import dataclass
from typing import List, Dict, Union
from flask import Flask, request, jsonify
import re

# ==== Type Definitions, feel free to add or modify ===========================
@dataclass
class CookbookEntry:
	name: str

@dataclass
class RequiredItem():
	name: str
	quantity: int

@dataclass
class Recipe(CookbookEntry):
	required_items: List[RequiredItem]

@dataclass
class Ingredient(CookbookEntry):
	cook_time: int


# =============================================================================
# ==== HTTP Endpoint Stubs ====================================================
# =============================================================================
app = Flask(__name__)

# Store your recipes here!
cookbook = {}

# Task 1 helper (don't touch)
@app.route("/parse", methods=['POST'])
def parse():
	data = request.get_json()
	recipe_name = data.get('input', '')
	parsed_name = parse_handwriting(recipe_name)
	if parsed_name is None:
		return 'Invalid recipe name', 400
	return jsonify({'msg': parsed_name}), 200

# [TASK 1] ====================================================================
# Takes in a recipeName and returns it in a form that
def parse_handwriting(recipeName: str) -> Union[str | None]:
	recipeName = recipeName.strip().replace('-', ' ').replace('_', ' ')
	recipeName = re.sub(r'[ ]+', ' ', recipeName)
	recipeName = re.sub(r'[^a-zA-Z\s]', '', recipeName)
	recipeName = recipeName.title()

	return recipeName if len(recipeName) > 0 else None


# [TASK 2] ====================================================================
# Endpoint that adds a CookbookEntry to your magical cookbook
@app.route('/entry', methods=['POST'])
def create_entry():
	data = request.get_json()
	if data['type'] not in ['recipe', 'ingredient']:
		return 'Wrong type', 400

	if data['type'] == 'ingredient' and data['cookTime'] < 0:
		return 'Negative cook time', 400

	if data['name'] in cookbook:
		return 'Name not unique', 400

	if data['type'] == 'recipe':
		seen = set()
		required_items = []
		for item in data['requiredItems']:
			if item['name'] in seen:
				return 'Required item not unique', 400
			seen.add(item['name'])
			required_items.append(RequiredItem(name=item['name'], quantity=item['quantity']))
		cookbook[data['name']] = Recipe(data['name'], required_items)
	elif data['type'] == 'ingredient':
		cookbook[data['name']] = Ingredient(name=data['name'], cook_time=data['cookTime'])

	return jsonify({}), 200


# [TASK 3] ====================================================================
# Endpoint that returns a summary of a recipe that corresponds to a query name
@app.route('/summary', methods=['GET'])
def summary():
	data = request.args.get('name')

	if not data or data not in cookbook:
		return 'Recipe not found', 400

	root = cookbook[data]
	if not isinstance(root, Recipe):
		return 'Not a recipe', 400

	final = {}
	final['name'] = root.name
	final['cookTime'] = 0
	final['ingredients'] = []
	final['seen'] = set()

	if not helper(root.required_items, 1, final):
		return 'Ingredient not found', 400

	return jsonify({
		"name": final['name'],
		"cookTime": final['cookTime'],
		"ingredients": final['ingredients']
	}), 200


def helper(required_items, multiplier, final):
	for child in required_items:
		if child.name not in cookbook:
			return False
		if isinstance(cookbook[child.name], Ingredient):
			if child.name in final['seen']:
				for item in final['ingredients']:
					if item['name'] == child.name:
						item['quantity'] = item['quantity'] + child.quantity * multiplier
						break
			else:
				final['ingredients'].append({"name": child.name, "quantity": child.quantity * multiplier})
				final['seen'].add(child.name)
			final['cookTime'] = final['cookTime'] + cookbook[child.name].cook_time * child.quantity * multiplier
		else:
			if not helper(cookbook[child.name].required_items, child.quantity * multiplier, final):
				return False
	return True

# =============================================================================
# ==== DO NOT TOUCH ===========================================================
# =============================================================================

if __name__ == '__main__':
	app.run(debug=True, port=8080)
